Content.Server/_Harmony/JoinQueue/JoinQueueManager.cs

```diff
@@ -0,0 +1,169 @@
+using System.Linq;
+using Content.Server.Connection;
+using Content.Shared.CCVar;
+using Content.Shared._Harmony.Common.JoinQueue;
+using Prometheus;
+using Robust.Server.Player;
+using Robust.Shared.Configuration;
+using Robust.Shared.Enums;
+using Robust.Shared.Network;
+using Robust.Shared.Player;
+using Robust.Shared.Timing;
+using Content.Shared._Harmony.CCVars;
+
+namespace Content.Server._Harmony.JoinQueue;
+
+/// <summary>
+///     Manages new player connections when the server is full and queues them up, granting access when a slot becomes free
+/// </summary>
+public sealed class JoinQueueManager : IJoinQueueManager
+{
+    private static readonly Gauge QueueCount = Metrics.CreateGauge(
+        "join_queue_total_count",
+        "Amount of players in queue.");
+
+    private static readonly Histogram QueueTimings = Metrics.CreateHistogram(
+        "join_queue_timings",
+        "Timings of players in queue",
+        new HistogramConfiguration()
+        {
+            LabelNames = new[] { "type" },
+            Buckets = Histogram.ExponentialBuckets(1, 2, 14),
+        });
+
+
+    [Dependency] private readonly IPlayerManager _player = default!;
+    [Dependency] private readonly IConfigurationManager _configuration = default!;
+    [Dependency] private readonly IServerNetManager _net = default!;
+    [Dependency] private readonly IConnectionManager _connection = default!;
+
+    /// <summary>
+    /// Queue of active player sessions
+    /// </summary>
+    private readonly List<ICommonSession> _queue = new();
+
+    private bool _isEnabled = false;
+
+    public int PlayerInQueueCount => _queue.Count;
+    public int ActualPlayersCount => _player.PlayerCount - PlayerInQueueCount; // Now it's only real value with actual players count that in game
+
+
+    public void Initialize()
+    {
+        _net.RegisterNetMessage<QueueUpdateMessage>();
+
+        _configuration.OnValueChanged(HCCVars.EnableQueue, OnQueueCVarChanged, true);
+        _player.PlayerStatusChanged += OnPlayerStatusChanged;
+    }
+
+
+    private void OnQueueCVarChanged(bool value)
+    {
+        _isEnabled = value;
+
+        if (!value)
+        {
+            foreach (var session in _queue)
+                session.Channel.Disconnect("Queue was disabled");
+        }
+    }
+
+
+    private async void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+    {
+        if (e.NewStatus == SessionStatus.Disconnected)
+        {
+            var wasInQueue = _queue.Remove(e.Session);
+
+            if (!wasInQueue && e.OldStatus != SessionStatus.InGame) // Process queue only if player disconnected from InGame or from queue
```

Comment:
According to the comment, this should be an `or` not an `and`
```suggestion
            if (!wasInQueue || e.OldStatus != SessionStatus.InGame) // Process queue only if player disconnected from InGame or from queue
```